components/codec/src/buffer.rs

```diff
@@ -15,21 +19,48 @@ pub trait BufferReader {
     ///
     /// This function may panic in some implementors when remaining space
     /// is not large enough to advance.
+    ///
+    /// TODO: We should make the panic behaviour deterministic.
     fn advance(&mut self, count: usize);
+
+    /// Read next several bytes as a slice and advance the position of internal cursor.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::Io` if there is not enough space to read specified number of bytes.
+    fn read_bytes(&mut self, count: usize) -> Result<&[u8]>;
 }
 
 impl<T: AsRef<[u8]>> BufferReader for std::io::Cursor<T> {
+    #[inline]
     fn bytes(&self) -> &[u8] {
         let pos = self.position() as usize;
         let slice = self.get_ref().as_ref();
         slice.get(pos..).unwrap_or(&[])
     }
 
+    #[inline]
     fn advance(&mut self, count: usize) {
         let mut pos = self.position();
         pos += count as u64;
         self.set_position(pos);
     }
+
+    fn read_bytes(&mut self, count: usize) -> Result<&[u8]> {
```

Comment:
I hope you wrote comments dissuading people from wasting time simplifying the function. I was beaten by the borrow checker. :(

Reply:
`// Do not ever try to waste time on simplifying this function. Here are failed challengers: xxxx`  ? 😄